aws/product/awsec2product.go

```diff
@@ -0,0 +1,179 @@
+package product
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/trackit/jsonlog"
+
+	"github.com/trackit/trackit2/config"
+	"github.com/trackit/trackit2/db"
+	"github.com/trackit/trackit2/models"
+)
+
+type (
+	PricingSku string
+
+	PricingAttribute struct {
+		InstanceType       string
+		CurrentGeneration  string
+		Vcpu               string
+		Memory             string
+		Storage            string
+		NetworkPerformance string
+		Tenancy            string
+		OperatingSystem    string
+		Ecu                string
+		Location           string
+		LocationType       string
+	}
+
+	PricingProduct struct {
+		Sku           PricingSku
+		ProductFamily string
+		Attributes    PricingAttribute
+	}
+
+	EC2Pricing struct {
+		Products map[PricingSku]PricingProduct
+	}
+)
+
+func storeRegionID(region string, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger, sqlTx models.XODB) error {
+	dbAwsRegion, err := models.AwsRegionByPretty(sqlTx, region)
+	if err != nil && err.Error() == "sql: no rows in result set" {
+		var newDbAwsRegion models.AwsRegion
+		newDbAwsRegion.Pretty = region
+		newDbAwsRegion.Region = region
+		if err := newDbAwsRegion.Insert(sqlTx); err != nil {
+			logger.Error("Error during dbAwsRegion inserting", err.Error())
+			return err
+		}
+		dbAwsProduct.RegionID = newDbAwsRegion.ID
+	} else if err == nil {
+		dbAwsProduct.RegionID = dbAwsRegion.ID
+	} else {
+		logger.Error("Error during dbAwsRegion fetching", err.Error())
+		return err
+	}
+	return nil
+}
+
+// storeAttributes stores all the attributes from PricingAttribute
+// to models.AwsProductEc2
+func storeAttributes(attributes *PricingAttribute, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger, sqlTx models.XODB) (err error) {
+	dbAwsProduct.InstanceType = attributes.InstanceType
+	if attributes.CurrentGeneration == "Yes" {
+		dbAwsProduct.CurrentGeneration = 1
+	}
+	if val, err := strconv.Atoi(attributes.Vcpu); err == nil {
+		dbAwsProduct.Vcpu = val
+	}
+	dbAwsProduct.Memory = attributes.Memory
+	dbAwsProduct.Storage = attributes.Storage
+	dbAwsProduct.NetworkPerformance = attributes.NetworkPerformance
+	dbAwsProduct.Tenancy = attributes.Tenancy
+	dbAwsProduct.OperatingSystem = attributes.OperatingSystem
+	if val, err := strconv.Atoi(attributes.Ecu); err == nil {
+		dbAwsProduct.Ecu = val
+	}
+	if attributes.LocationType == "AWS Region" {
+		err = storeRegionID(attributes.Location, dbAwsProduct, logger, sqlTx)
+	}
+	return err
+}
+
+// importResult parses the body returned by downloadJSON and
+// inserts the pricing in the database.
+func importResult(reader io.ReadCloser, logger jsonlog.Logger, sqlTx models.XODB) error {
+	defer reader.Close()
+
+	decoder := json.NewDecoder(reader)
+	var bodyMap EC2Pricing
+	if err := decoder.Decode(&bodyMap); err != nil {
```

Comment:
Please stream the JSON document as per my previous review. This document is known to be quite big and reasonable memory footprint is a requirement for this project.